### PR TITLE
Content type mapping for markdown

### PR DESCRIPTION
**Pull request by korroz**
Since it is now officially registered.

See: https://tools.ietf.org/html/rfc7763

### CODE REVIEW

**Comment by Tratcher** [comment]
Wouldn't sites prefer to render markdown rather than serve it directly?

**Comment by davidfowl** [comment]
Yes, I'm also confused by this PR

**Comment by korroz** [comment]
Maybe _sites_ would prefer that, but if you use `StaticFiles` why wouldn't you expect it to serve markdown as a known content type?
Especially now that it's registered as well.

Hm, maybe I'm missing the point here because I don't see how the server side rendering scenario is relevant to `StaticFiles`.

**Comment by korroz** [comment]
Okay, I've thought a bit more about the consequences. The usage of `StaticFiles` seems to roughly be these two scenarios:

 1. Serving `wwwroot`. As in the VS and yeoman templates.
   - This case is the default, which most people will encounter. So the question is, if you put a markdown file in `wwwroot` do you expect it to be served?
 2. Serving application specific paths.
   - Here a developer is telling his application specifically to serve files from a specified location. Again the question is would you expect markdown files to be served?

I would still say yes to both those scenarios.

There might be a case for when a developer wants to implement serverside rendering of markdown, but how to anticipate potential problems with such a scenario seems quite out of scope of `StaticFiles`. Even so, lets examine what could happen hypothetically.

Here are two options I could imagine:

 1. Use MVC.
   - Either you'd use some sort of markdown partial or implement a markdown viewengine of some sort. In both of these cases you would not want to store the files in `wwwroot` (just like views aren't stored there) so conflicts seems unlikely.
 2. Implement a middleware
   - You would quickly realize the precendence of this middleware in relation to any usage of `StaticFiles`. And again, `wwwroot` would not be the place for the files as per the recommendations.

Another possibility I could think of for not accepting this PR is that the default mappings have been decided to only mirror the defaults of some other software, like out of the box IIS or something. Which is valid if that's the case.

The last reason I can think of to reject the PR is that the markdown registration is so recent. And that markdown itself isn't widespread enough to validate it's inclusion in the default mappings.

I think this lays it all out plainly for all to see, but of course there might be things I just can't see clearly. I've done my best to anticipate all concerns though.

Are we still confused?

**Comment by Tratcher** [comment]
You make good points on the location of the files and the expectation to serve them. I think the biggest question is why do you want to serve them in their raw format? And how common a need this is?

We expect lots of developers to need to serve uncommon file types so we made the list editable in code. Does that list need to be easier to modify?

Yes we did copy the default mappings from IIS, but that doesn't mean we need to keep them in sync.

**Comment by korroz** [comment]
In my particular case I'm deferring rendering of markdown to the frontend. Not only are there some excellent javascript libraries out there that can handle rendering, but it is also useful to have the source for features such as editing.

> How common would this be?

I really can't say. I have no real data on this. Most markdown file usage I think is in code repositories. Most of which let their repo hosting solution do the rendering. For those that want to get more serious about using markdown, static site generators seems like the way to go. Although there are also browser extensions for rendering markdown files being served raw, which indicates that such things already do occur.

> Is the mapping list hard to modify?

Not at all. The API is already nice and easy to use if you need to serve arbitrary files that aren't covered by default. The following few lines gets the job done.

```cs
// Map markdown file extension
var contentTypeProvider = new FileExtensionContentTypeProvider();
contentTypeProvider.Mappings.Add(".md", "text/markdown");
app.UseStaticFiles(new StaticFileOptions { ContentTypeProvider = contentTypeProvider });
```

I suppose the API could still be improved if one were to give it a lot of thought. But that's not the issue.

> Is it worth adding?

I know one should always be cautious about adding defaults. I don't know how common this scenario is either, although I do find it likely it won't be zero. But even if I did have good data on how common this would be the question would then be how common is common enough?
Even if it is very uncommon this addition to the defaults would make it a non-issue, it would just work. The downside of adding `text/markdown` to the defaults is that people might unintentionally serve their markdown files. What we should be comparing is how likely _that_ scenario is compared to the scenario of being forced to add the lines of code above. I find it extremely unlikely that anyone would be surprised that putting `.md` files into `wwwroot` would actually serve them, while finding out that they unexpectedly aren't served seems more likely.

Gosh.. didn't think I'd have to analyse this addition so deeply. Hehe.

Contrary to how it might seem, my opinion about this isn't that strong, I just saw something that caused friction while using ASP.NET and looked deeper into the issue, saw that markdown had been officially registered with the IETF and thought I'd help out with a super tiny contribution.

The only additional arguments that could be made is how the usefulness of serving markdown compares with the existing defaults. But they have the benefit of being legacy which is hard to argue against. And also I don't really want to go there. So I think I've exhausted my thoughts on this now and have nothing left to add, so the rest is up you. I'll just repeat my strongest argument, comparing the downside of accepting this PR with the implications of declining it, although I have no data to back it up.

 - Accept
   - I find it extremely unlikely that anyone would be surprised that putting `.md` files into `wwwroot` would actually serve them.
 - Decline
   - Finding out that `.md` files unexpectedly aren't served seems more likely.
 - TL;DR
   - Accepting seems to cause less problems in the future. But, of course, it's really up to you.

**Comment by Tratcher** [comment]
@DamianEdwards This is a PM call. We need official guild-lines somewhere for when we'll add entries to StaticFile's default content-type list.

**Comment by Tratcher** [comment]
@DamianEdwards approved for post RC2 and we will revisit the criteria question as https://github.com/aspnet/StaticFiles/issues/129.
Rebased and merged to dev. Thanks for indulging our philosophical discussions.

**Comment by korroz** [comment]
@Tratcher No problem, was fun to contribute!
